File: src/data/o1/Find_Beautiful_Indices_in_the_Given_Array_II/Find_Beautiful_Indices_in_the_Given_Array_II.java
import java.util.*;

class Solution {
    public List<Integer> beautifulIndices(String s, String a, String b, int k) {
        List<Integer> result = new ArrayList<>();
        List<Integer> aList = findOccurrences(s, a);
        List<Integer> bList = findOccurrences(s, b);
        
        if(aList.isEmpty() || bList.isEmpty()) return result;
        
        int j = 0;
        int bSize = bList.size();
        
        for(int i : aList){
            // Move j to the first bIndex >= i - k
            while(j < bSize && bList.get(j) < i - k){
                j++;
            }
            // Check if current bList[j] is within i + k
            if(j < bSize && Math.abs(bList.get(j) - i) <= k){
                result.add(i);
                continue;
            }
            // Check the previous bList[j-1] if j > 0
            if(j > 0 && Math.abs(bList.get(j-1) - i) <= k){
                result.add(i);
            }
        }
        
        return result;
    }
    
    private List<Integer> findOccurrences(String s, String pattern){
        List<Integer> list = new ArrayList<>();
        int n = s.length();
        int m = pattern.length();
        for(int i = 0; i <= n - m; i++){
            if(s.substring(i, i + m).equals(pattern)){
                list.add(i);
            }
        }
        return list;
    }
}

public class Find_Beautiful_Indices_in_the_Given_Array_II {
public static void runTests(Solution solution) {
    // Test case 1: Simple case with one beautiful index
    String s1 = "abcde";
    String a1 = "ab";
    String b1 = "de";
    int k1 = 4;
    List<Integer> expected_result1 = Arrays.asList(0);
    assert solution.beautifulIndices(s1, a1, b1, k1).equals(expected_result1) : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Multiple beautiful indices
    String s2 = "ababab";
    String a2 = "ab";
    String b2 = "ba";
    int k2 = 1;
    List<Integer> expected_result2 = Arrays.asList(0, 2, 4);
    assert solution.beautifulIndices(s2, a2, b2, k2).equals(expected_result2) : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: No beautiful indices
    String s3 = "isawsquirrelnearmysquirrelhouseohmy";
    String a3 = "my";
    String b3 = "squirrel";
    int k3 = 15;
    List<Integer> expected_result3 = Arrays.asList(16, 33);
    assert solution.beautifulIndices(s3, a3, b3, k3).equals(expected_result3) : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Overlapping matches with larger k
    String s4 = "abababab";
    String a4 = "ab";
    String b4 = "ab";
    int k4 = 2;
    List<Integer> expected_result4 = Arrays.asList(0, 2, 4, 6);
    assert solution.beautifulIndices(s4, a4, b4, k4).equals(expected_result4) : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Only one match each for a and b, within range k
    String s5 = "abcdef";
    String a5 = "abc";
    String b5 = "def";
    int k5 = 5;
    List<Integer> expected_result5 = Arrays.asList(0);
    assert solution.beautifulIndices(s5, a5, b5, k5).equals(expected_result5) : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Overlapping b close to the end of a with k = 0
    String s6 = "abcd";
    String a6 = "a";
    String b6 = "a";
    int k6 = 4;
    List<Integer> expected_result6 = Arrays.asList(0);
    assert solution.beautifulIndices(s6, a6, b6, k6).equals(expected_result6) : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Non-overlapping multiple beautiful indices with exact k distance
    String s7 = "aaaaaa";
    String a7 = "aa";
    String b7 = "aa";
    int k7 = 2;
    List<Integer> expected_result7 = Arrays.asList(0, 1, 2, 3, 4);
    assert solution.beautifulIndices(s7, a7, b7, k7).equals(expected_result7) : "Test case 7 failed. Expected " + expected_result7;
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 3000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/o1/Find_Beautiful_Indices_in_the_Given_Array_II/Find_Beautiful_Indices_in_the_Given_Array_II.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> beautifulIndices(string s, string a, string b, int k) {
        vector<int> a_indices;
        vector<int> b_indices;
        int n = s.length();
        int len_a = a.length();
        int len_b = b.length();
        
        // Find all indices where a appears
        for(int i = 0; i <= n - len_a; ++i){
            if(s.substr(i, len_a) == a){
                a_indices.push_back(i);
            }
        }
        
        // Find all indices where b appears
        for(int i = 0; i <= n - len_b; ++i){
            if(s.substr(i, len_b) == b){
                b_indices.push_back(i);
            }
        }
        
        vector<int> result;
        int m = b_indices.size();
        if(m == 0){
            // No b found, no beautiful indices
            return result;
        }
        
        // Iterate through a_indices and find if any b_indices within [i -k, i +k]
        for(auto &i_idx : a_indices){
            // Find the first b_idx >= i_idx - k
            int j = lower_bound(b_indices.begin(), b_indices.end(), i_idx - k) - b_indices.begin();
            // Check if any b_idx in [i_idx -k, i_idx +k]
            if(j < m && b_indices[j] <= i_idx + k){
                result.push_back(i_idx);
            }
        }
        
        return result;
    }
};


//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple case with one beautiful index
    std::string s = "abcde";
    std::string a = "ab";
    std::string b = "de";
    int k = 4;
    std::vector<int> expected_result = {0};
    assert(solution.beautifulIndices(s, a, b, k) == expected_result && "Test case 1 failed.");

    // Test case 2: Multiple beautiful indices
    s = "ababab";
    a = "ab";
    b = "ba";
    k = 1;
    expected_result = {0, 2, 4};
    assert(solution.beautifulIndices(s, a, b, k) == expected_result && "Test case 2 failed.");

    // Test case 3: No beautiful indices
    s = "isawsquirrelnearmysquirrelhouseohmy";
    a = "my";
    b = "squirrel";
    k = 15;
    expected_result = {16, 33};
    assert(solution.beautifulIndices(s, a, b, k) == expected_result && "Test case 3 failed.");

    // Test case 4: Overlapping matches with larger k
    s = "abababab";
    a = "ab";
    b = "ab";
    k = 2;
    expected_result = {0, 2, 4, 6};
    assert(solution.beautifulIndices(s, a, b, k) == expected_result && "Test case 4 failed.");

    // Test case 5: Only one match each for a and b, within range k
    s = "abcdef";
    a = "abc";
    b = "def";
    k = 5;
    expected_result = {0};
    assert(solution.beautifulIndices(s, a, b, k) == expected_result && "Test case 5 failed.");

    // Test case 6: Overlapping b close to the end of a with k = 0
    s = "abcd";
    a = "a";
    b = "a";
    k = 4;
    expected_result = {0};
    assert(solution.beautifulIndices(s, a, b, k) == expected_result && "Test case 6 failed.");

    // Test case 7: Non-overlapping multiple beautiful indices with exact k distance
    s = "aaaaaa";
    a = "aa";
    b = "aa";
    k = 2;
    expected_result = {0, 1, 2, 3, 4};
    assert(solution.beautifulIndices(s, a, b, k) == expected_result && "Test case 7 failed.");
}


int main() {
    Solution solution;
    for (int i = 0; i < 163840; ++i) {
        testSolution(solution);  
    }
    return 0;
}
